internal/secretsource/url/url.go

```diff
@@ -0,0 +1,401 @@
+	RetryBackoff types.NullDuration `json:"retryBackoff"`
+}
+
+// newConfig creates a new extConfig instance with default values.
+func newConfig() extConfig {
+	return extConfig{
+		Method:                 http.MethodGet,                               // GET method
+		RequestsPerMinuteLimit: null.NewInt(300, false),                      // 300 requests per minute
+		RequestsBurst:          null.NewInt(10, false),                       // Allow a burst of 10 requests
+		Timeout:                types.NewNullDuration(30*time.Second, false), // 30 seconds timeout
+		MaxRetries:             null.NewInt(3, false),                        // 3 retry attempts
+		RetryBackoff:           types.NewNullDuration(1*time.Second, false),  // 1 second base backoff
+	}
+}
+
+//nolint:gochecknoinits // This is how k6 secret source registration works.
+func init() {
+	secretsource.RegisterExtension("url", func(params secretsource.Params) (secretsource.Source, error) {
+		config, err := getConfig(params.ConfigArgument, params.FS)
+		if err != nil {
+			return nil, fmt.Errorf("missing or invalid config: %w", err)
+		}
+
+		return &urlSecrets{
+			config: config,
+			httpClient: &http.Client{
+				Timeout: time.Duration(config.Timeout.Duration),
+			},
+			limiter: newLimiter(int(config.RequestsPerMinuteLimit.Int64), int(config.RequestsBurst.Int64)),
+		}, nil
+	})
+}
+
+type urlSecrets struct {
+	config     extConfig
+	httpClient *http.Client
+	limiter    limiter
+}
+
+func (us *urlSecrets) Description() string {
+	return "URL-based secret source"
+}
+
+func (us *urlSecrets) Get(key string) (string, error) {
+	ctx := context.Background()
+
+	if err := us.limiter.Wait(ctx); err != nil {
+		return "", fmt.Errorf("rate limiter error: %w", err)
+	}
+
+	var secret string
+	// MaxRetries is the number of retries, so total attempts = retries + 1
+	maxAttempts := int(us.config.MaxRetries.Int64) + 1
+	backoff := time.Duration(us.config.RetryBackoff.Duration)
+
+	err := retry(ctx, maxAttempts, backoff, func() error {
+		// Replace {key} placeholder in URL template
+		escapedKey := url.PathEscape(key)
+		url := strings.ReplaceAll(us.config.URLTemplate, "{key}", escapedKey)
+
+		req, err := http.NewRequestWithContext(ctx, us.config.Method, url, nil)
+		if err != nil {
+			return &retryableError{err: fmt.Errorf("failed to create request: %w", err), statusCode: 0}
+		}
+
+		// Add headers
+		for k, v := range us.config.Headers {
+			req.Header.Set(k, v)
+		}
+
+		response, err := us.httpClient.Do(req)
+		if err != nil {
+			// Network errors are retryable
+			return &retryableError{err: fmt.Errorf("failed to get secret: %w", err), statusCode: 0}
+		}
+		defer func() { _ = response.Body.Close() }()
+
+		if response.StatusCode != http.StatusOK {
+			statusErr := fmt.Errorf("status code %d: %w", response.StatusCode, errFailedToGetSecret)
+
+			// Check if this is a retryable status code
+			if isRetryableError(response.StatusCode, nil) {
```

Comment:
it makes no sense to always be sending `nil` and to be checking it in the funciton a